### ISSUE TITLE
Bug: str() and validator objects

### ISSUE THREAD

**Issue by AnnePetersen1** [data-cleaning/validate#81]
I get an error when I use the `str()` function on your `validator` objects:

``` 
library(validate)
> val <- validator(var1 < var2)
> class(val)
[1] "validator"
attr(,"package")
[1] "validate"
> str(val)
Error in x$rules[[i]] : 
  attempt to select less than one element in get1index
``` 

I'm using the latest version (0.2.0) from CRAN.

**Comment by markvanderloo** [comment]
Thanks for reporting!

We have not implemented a `str` method for objects of class `confrontation` or `validation` so the generic method is crashing here somehow.

**Comment by AnnePetersen1** [comment]
No problem. 

It's super weird, I have never seen `str()` crashing before.  I hope you can solve it so that it gets easier to inspect what's in the validation objects.

**Comment by markvanderloo** [comment]
This appears to be a collision between `str` and reference classes with an S4 method for `[[`. The following gives a similar error. It looks like `str` is using `[[` rather than `$` to access the fields of a Reference Class. This is an unsafe assumption since overloading of `[[` is clearly permitted by R.

```r
foo <- setRefClass("foo",fields=list(bar="character"))
setMethod("[[","foo",function(x,i,j,...,exact=TRUE){
  x$bar[[i]]
})

obj <- foo(bar=c("hello","dolly"))
obj[[2]]
str(obj)
```

**Comment by markvanderloo** [comment]
To see the contents of a `validator` object you can just print them, or use `[[` to print the contents for each rule precisely, e.g.

```r
v <- validator(x>0, y < x)
v[[1]]
```

**Comment by edwindj** [comment]
Seems to be solved by R-core in version 3.4.4

**Comment by talgalili** [comment]
Hooray!
